internal/clients/aws.go

```diff
@@ -6,62 +6,184 @@ package clients
+			},
+		}
+		account, err := getAccountId(ctx, c, mg)
 		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+			return terraform.Setup{}, errors.Wrap(err, "cannot get account id")
 		}
-		if cfg.Region == "" && mg.GetObjectKind().GroupVersionKind().Group == "iam.aws.upbound.io" {
-			cfg.Region = "us-east-1"
+		ps.ClientMetadata = map[string]string{
+			keyAccountId: account,
 		}
-		creds, err := cfg.Credentials.Retrieve(ctx)
+
+		if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+			err = DefaultTerraformSetupBuilder(ctx, c, mg, &ps)
+			if err != nil {
+				return terraform.Setup{}, errors.Wrap(err, "cannot build terraform configuration")
+			}
+		} else {
+			err = pushDownTerraformSetupBuilder(ctx, c, mg, pc, &ps)
+			if err != nil {
+				return terraform.Setup{}, errors.Wrap(err, "cannot build terraform configuration")
+			}
+		}
+		return ps, nil
+	}
+}
+
+func pushDownTerraformSetupBuilder(ctx context.Context, c client.Client, mg resource.Managed, pc *v1beta1.ProviderConfig, ps *terraform.Setup) error { //nolint:gocyclo
+	if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+		return errors.New("shared scheduler cannot be used because the length of assume role chain array " +
+			"is more than 1 or endpoint configuration is not nil")
+	}
+
+	cfg, err := getAWSConfig(ctx, c, mg)
+	if err != nil {
+		return errors.Wrap(err, "cannot get AWS config")
+	}
+	ps.Configuration = map[string]any{
+		keyRegion: cfg.Region,
+	}
+
+	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
+	case authKeyWebIdentity:
+		if pc.Spec.Credentials.WebIdentity == nil {
+			return errors.New(`spec.credentials.webIdentity of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		if pc.Spec.Credentials.WebIdentity.RoleARN == nil {
+			return errors.New(`spec.credentials.webIdentity.roleARN of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = map[string]any{
+			keyRoleArn:              pc.Spec.Credentials.WebIdentity.RoleARN,
+			keySessionName:          pc.Spec.Credentials.WebIdentity.RoleSessionName,
+			keyWebIdentityTokenFile: os.Getenv(envWebIdentityTokenFile),
+		}
+	case authKeyUpbound:
+		if pc.Spec.Credentials.Upbound == nil || pc.Spec.Credentials.Upbound.WebIdentity == nil {
+			return errors.New(`spec.credentials.upbound.webIdentity of ProviderConfig cannot be nil when the credential source is "Upbound"`)
+		}
+		if pc.Spec.Credentials.Upbound.WebIdentity.RoleARN == nil {
+			return errors.New(`spec.credentials.upbound.webIdentity.roleARN of ProviderConfig cannot be nil when the credential source is "Upbound"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = map[string]any{
+			keyRoleArn:              pc.Spec.Credentials.Upbound.WebIdentity.RoleARN,
+			keySessionName:          pc.Spec.Credentials.Upbound.WebIdentity.RoleSessionName,
```

Comment:
Did not try yet, but it should complain with the above error message given for `WebIdentity`.